File: packages/safe-bash/tests/compatibility/bash-strict-extension-independent-20260829/launch.py
import hashlib
import json
import os
from pathlib import Path
import selectors
import signal
import subprocess
import sys
import time
import traceback

OWN = Path("/Users/kjopek/Workspace/safe-bash/tests/compatibility/bash-strict-extension-independent-20260829")


def digest(filename):
    result = hashlib.sha256()
    with open(filename, 'rb') as stream:
        while block := stream.read(65536):
            result.update(block)
    return result.hexdigest()


output = OWN / 'actual-v1'
output.mkdir()
with open(output / 'events.jsonl', 'x', buffering=1) as events, open(output / 'stdout.raw', 'xb', buffering=0) as stdout, open(output / 'stderr.raw', 'xb', buffering=0) as stderr:
    record = {'primaryPresent': False, 'cleanupPresent': False, 'closed': False, 'signals': [], 'bytes': 0}
    child = None
    began = time.monotonic()
    try:
        events.write(json.dumps({'event': 'capture-before-admission', 'pid': os.getpid()}) + '\n')
        assert sys.argv[1:] == ['--run']
        seal = json.loads((OWN / 'PRESEAL.json').read_text())
        executor = json.loads((OWN / 'EXECUTOR.json').read_text())
        assert digest(seal['node']['path']) == seal['node']['sha256']
        for row in executor['files']:
            target = Path(seal['repo']) / row['path']
            assert target.is_file() and not target.is_symlink()
            assert target.stat().st_size == row['bytes'] and digest(target) == row['sha256'], row['path']
        command = [seal['node']['path'], str(OWN / 'run.mjs'), '--run']
        events.write(json.dumps({'event': 'enrolled-before-spawn', 'command': command}) + '\n')
        child = subprocess.Popen(command, cwd=seal['repo'], env={'PATH': str(Path(seal['node']['path']).parent), 'HOME': seal['workParent'], 'TMPDIR': seal['workParent']}, stdin=subprocess.DEVNULL, stdout=subprocess.PIPE, stderr=subprocess.PIPE, start_new_session=True)
        record['pid'] = child.pid
        watcher = selectors.DefaultSelector()
        watcher.register(child.stdout, selectors.EVENT_READ, stdout)
        watcher.register(child.stderr, selectors.EVENT_READ, stderr)
        while watcher.get_map():
            assert time.monotonic() - began < 3300, 'outer deadline'
            for key, mask in watcher.select(0.25):
                data = os.read(key.fileobj.fileno(), 65536)
                if not data:
                    watcher.unregister(key.fileobj)
                    key.fileobj.close()
                else:
                    record['bytes'] += len(data)
                    assert record['bytes'] <= 16 * 1024 * 1024, 'outer capture cap'
                    key.data.write(data)
        record['exitCode'] = child.wait(timeout=5)
        record['closed'] = True
        watcher.close()
        try:
            os.killpg(child.pid, 0)
            raise AssertionError('owned group remains after coordinator close')
        except ProcessLookupError:
            record['ownedGroupAbsent'] = True
    except BaseException:
        record['primaryPresent'] = True
        record['primary'] = traceback.format_exc()
        stderr.write(record['primary'].encode())
    finally:
        if child is not None:
            try:
                try:
                    os.killpg(child.pid, 0)
                except ProcessLookupError:
                    pass
                else:
                    record['signals'].append('SIGTERM')
                    os.killpg(child.pid, signal.SIGTERM)
                    try:
                        child.wait(timeout=2)
                    except subprocess.TimeoutExpired:
                        pass
                    try:
                        os.killpg(child.pid, 0)
                    except ProcessLookupError:
                        pass
                    else:
                        record['signals'].append('SIGKILL')
                        os.killpg(child.pid, signal.SIGKILL)
                    child.wait(timeout=3)
                if child.poll() is not None:
                    record['closed'] = True
                for stream in [child.stdout, child.stderr]:
                    if not stream.closed:
                        stream.close()
            except BaseException:
                record['cleanupPresent'] = True
                record['cleanup'] = traceback.format_exc()
        record['elapsedSeconds'] = time.monotonic() - began
        events.write(json.dumps({'event': 'terminal', **record}) + '\n')
        (output / 'TERMINAL.json').write_text(json.dumps(record, indent=2) + '\n')
    sys.exit(78 if record['primaryPresent'] or record['cleanupPresent'] or record['signals'] or not record['closed'] else record['exitCode'])

File: packages/safe-bash/tests/commands/node-worker-experiments-20260828/preparation-v9-l02-repair/launch.mjs
import * as fs from 'node:fs';
import { spawn } from 'node:child_process';
import { createHash } from 'node:crypto';
import { fileURLToPath } from 'node:url';
import { createCaptureOwner } from '../preparation-v7-capture/capture-owner.mjs';
const root = new URL('.', import.meta.url);
const digest = bytes => createHash('sha256').update(bytes).digest('hex');
function bounded(url, maximum) {
  const filename = fileURLToPath(url), stat = fs.lstatSync(filename);
  if (!stat.isFile() || stat.isSymbolicLink() || fs.realpathSync(filename) !== filename || stat.size > maximum) throw Error('regular bounded input');
  return fs.readFileSync(filename);
}
export async function run(token) {
  const output = new URL('run-01/', root);
  fs.mkdirSync(output);
  const capture = createCaptureOwner({ open: url => fs.openSync(url,'wx',0o600), write: (fd,bytes,offset,length) => fs.writeSync(fd,bytes,offset,length), close: fd => fs.closeSync(fd) });
  if (!capture.acquire([new URL('stdout.raw',output),new URL('stderr.raw',output)])) return { capture:capture.snapshot(), launched:false };
  let child, timer, killTimer, code = null, signal = null, childErrorPresent = false, closed = false;
  const started = Date.now();
  try {
    const sealBytes = bounded(new URL('PRESEAL.json',root),65536);
    if (digest(sealBytes) !== token) throw Error('preseal token');
    const seal = JSON.parse(sealBytes);
    for (const record of [...seal.modules,...seal.hostInputs]) {
      const bytes = bounded(new URL(record.path,root),65536);
      if (bytes.length !== record.bytes || digest(bytes) !== record.sha256) throw Error('input hash');
    }
    const nodeStat = fs.lstatSync(seal.node.path);
    if (!nodeStat.isFile() || nodeStat.isSymbolicLink() || fs.realpathSync(seal.node.path) !== seal.node.path || nodeStat.size !== seal.node.bytes) throw Error('Node admission');
    const nodeHash = createHash('sha256');
    for await (const bytes of fs.createReadStream(seal.node.path,{highWaterMark:65536})) nodeHash.update(bytes);
    if (nodeHash.digest('hex') !== seal.node.sha256) throw Error('Node hash');
    const args = [...seal.argv];
    await new Promise(resolve => {
      const stop = () => { if (child && !closed) { child.kill('SIGTERM'); killTimer ??= setTimeout(() => { if (!closed) child.kill('SIGKILL'); },2000); } };
      timer = setTimeout(() => { capture.record(undefined,'outer-deadline'); stop(); },30000);
      try {
        child = spawn(seal.node.path,args,{cwd:fileURLToPath(root),env:seal.env,stdio:['ignore','pipe','pipe']});
      } catch (value) { capture.record(value,'spawn'); clearTimeout(timer); resolve(); return; }
      child.once('close',(exitCode,exitSignal)=>{ closed=true;code=exitCode;signal=exitSignal;clearTimeout(timer);clearTimeout(killTimer);resolve(); });
      child.once('error',value=>{ childErrorPresent=true;capture.record(value,'child-error');stop(); });
      child.stdout.on('data',bytes=>{ if(!capture.write(0,bytes))stop(); });
      child.stderr.on('data',bytes=>{ if(!capture.write(1,bytes))stop(); });
      child.stdout.on('error',value=>{ capture.record(value,'stdout-error');stop(); });
      child.stderr.on('error',value=>{ capture.record(value,'stderr-error');stop(); });
    });
  } catch (value) { capture.record(value,'outer-admission'); }
  finally { clearTimeout(timer);clearTimeout(killTimer);capture.close(); }
  return { schema:'parent-only-owner-v1', launched:!!child, pid:child?.pid??null, closed, code, signal, childErrorPresent, elapsedMs:Date.now()-started, capture:capture.snapshot(), primary:capture.primary, secondary:capture.secondary };
}
